### PR TITLE
docker-compose.yml and Dockerfile for launch.sh

### LINKED ISSUE #50: Fix launching docker and services at the same time

### PR DESCRIPTION
**Pull request by YaroslavKSE**
Created a sepatate folder for laucnh srpits for all microservices and frontend. Created a separate launch.sh script in the root directory which runs all scripts form LaunchScripts. Works perfectly!

Closes #50 #51

### CODE REVIEW

File: launch.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
+
+# Start Docker Compose for PostgreSQL
+echo "Starting PostgreSQL for Identity Service"
+cd Identity
+docker-compose up -d
+cd ..
+
+# Wait for PostgreSQL to be ready (Optional, you might need a more sophisticated wait mechanism)
+sleep 3
```
Comment (roflmyrlok) [inline launch.sh:10]: use until: try if db ready; sleep 1

```bash
until pg_isready -h localhost -p 5432 -U myuser; do
  echo "Waiting for PostgreSQL to be ready..."
  sleep 1
done
echo "PostgreSQL is ready!"

File: launch.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+cd Identity
+export ASPNETCORE_URLS=https://localhost:7108
+export ASPNETCORE_ENVIRONMENT=Development
+dotnet restore Identity.sln
+dotnet run Identity.sln
+
+
+
+
+echo "Starting CourseTemplate Service"
```
Comment (roflmyrlok) [inline launch.sh:28]: Separate launch of each solution to separate file? It is to complicated to call and run each solution from the same file.

**Review by roflmyrlok — CHANGES_REQUESTED** [review]

File: launch.sh (hunk 1 above)
Comment (YaroslavKSE) [reply launch.sh]: Fixed in recent commit

File: launch.sh (hunk 2 above)
Comment (YaroslavKSE) [reply launch.sh]: Fixed in recent commit

File: launch.sh (hunk 1 above)
Comment (ykhadnevich) [reply launch.sh]: Nothing was fixed

File: LaunchScripts/launchMaterials.sh [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+# Start Docker Compose for PostgreSQL
+echo "Starting PostgreSQL for Materials Service"
+cd ..
+cd Materials
+docker-compose up -d
+
+# Wait for PostgreSQL to be ready
+sleep 3
```
Comment (Bardin08) [inline LaunchScripts/launchMaterials.sh:10]: Sleep is not a reliable way. Try to use the script listed below. Also, you can reuse it for all services without duplicating this logic (at least to await required containers).

> [!IMPORTANT]
> You have to pass the container name as a parameter value

```sh
#!/bin/bash

# ANSI escape code for colors
CYAN="\033[36m"
RED="\033[31m"
GREEN="\033[32m"

RESET="\033[0m"

if [ $# -ne 1 ]; then
    echo "Usage: $0 <container-name>"
    exit 1
fi

container_name="$1"

echo "Starting Docker Container"
docker-compose up -d

timeout=$((SECONDS + 60))  # Set a timeout of 60 seconds
interval=1

while [ $SECONDS -lt $timeout ]; do
    # Check if PostgreSQL container is running
    if docker ps | grep -q "$container_name"; then
        echo -e "${GREEN} Container is running! ${RESET}"
        break
    else
        # Get current timestamp in [hh:mm:ss] format
        timestamp=$(date "+[%H:%M:%S]")
        echo -e "${CYAN}${timestamp} ${container_name} is not running. Retrying in $interval seconds... ${RESET}"
        sleep $interval
    fi
done

if [ $SECONDS -ge $timeout ]; then
    echo -e "${RED} ${container_name} did not start within a minute. Aborting. ${RESET}"
    exit 1
fi

exit 0
```

Usage:

```sh
./launch.sh postgres_1234
```

File: LaunchScripts/launchIdentity.sh [hunk 4]
```diff
@@ -1,7 +1,6 @@
 #!/bin/bash
 
 # Start Docker Compose for PostgreSQL
-
 echo "Starting PostgreSQL for Identity Service"
 cd Identity
 docker-compose up -d
```
Comment (Bardin08) [inline LaunchScripts/launchIdentity.sh:6]: Add container awaiting script from another comment

File: LaunchScripts/launchCourseTemplate.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+# Start Docker Compose for PostgreSQL
+echo "Starting PostgreSQL for CourseTemplate Service"
+cd ..
+cd CourseTemplate
+docker-compose up -d
+
+# Wait for PostgreSQL to be ready
+
+sleep 3
```
Comment (Bardin08) [inline LaunchScripts/launchCourseTemplate.sh:11]: Add container awaiting script from another comment

File: LaunchScripts/launchCourseStream.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+# Start Docker Compose for PostgreSQL
+echo "Starting PostgreSQL for CourseStream Service"
+cd ..
+cd CourseStream
+docker-compose up -d
+
+sleep 3
```
Comment (Bardin08) [inline LaunchScripts/launchCourseStream.sh:9]: Add container awaiting script from another comment

File: frontend/.idea/.idea.frontend.dir/.idea/.gitignore
Comment (Bardin08) [inline frontend/.idea/.idea.frontend.dir/.idea/.gitignore:1]: Remove Rider's cache

**Review by Bardin08 — CHANGES_REQUESTED** [review]
Take a look at the script I attached to one of the comments. Try to figure out how to use it for all services and how using it you can create reliable & simple postgreSQL container awaiter.

Also remove Rider's cache from the PR!

File: LaunchScripts/launchMaterials.sh (hunk 3 above)
Comment (YaroslavKSE) [reply LaunchScripts/launchMaterials.sh]: Made improvements

File: LaunchScripts/launchIdentity.sh (hunk 4 above)
Comment (YaroslavKSE) [reply LaunchScripts/launchIdentity.sh]: Made improvements

File: LaunchScripts/launchCourseTemplate.sh (hunk 5 above)
Comment (YaroslavKSE) [reply LaunchScripts/launchCourseTemplate.sh]: Made improvements

File: LaunchScripts/launchCourseStream.sh (hunk 6 above)
Comment (YaroslavKSE) [reply LaunchScripts/launchCourseStream.sh]: Made improvements

**Pull request closed without merging**